File: src/main/java/dev/kensa/state/CapturedInteractionBuilder.java
package dev.kensa.state;

import dev.kensa.util.Attributes;

public final class CapturedInteractionBuilder {

    private static final String TEMPLATE = "%s __idx __from %s to %s";

    public static CapturedInteractionBuilder from(Party party) {
        return new CapturedInteractionBuilder(party);
    }

    private Attributes attributes = Attributes.of();
    private Party toParty;
    private String group;
    private Object content;
    private String contentDescriptor;
    private final Party fromParty;

    private CapturedInteractionBuilder(Party party) {
        fromParty = party;
    }

    public CapturedInteractionBuilder to(Party toParty) {
        this.toParty = toParty;

        return this;
    }

    public CapturedInteractionBuilder group(String group) {
        this.group = group;

        return this;
    }

    public CapturedInteractionBuilder underTest(boolean isUnderTest) {
        if (group == null & isUnderTest) {
            group = "Test";
        }

        return this;
    }

    public CapturedInteractionBuilder with(Object content, String contentDescriptor) {
        this.content = content;
        this.contentDescriptor = contentDescriptor;

        return this;
    }

    public CapturedInteractionBuilder with(Attributes attributes) {
        this.attributes = attributes;

        return this;
    }

    void applyTo(CapturedInteractions capturedInteractions) {
        String keyPrefix = group == null ? "" : String.format("(%s) ", group);

        String key = keyPrefix + String.format(TEMPLATE, contentDescriptor, fromParty.asString(), toParty.asString());

        capturedInteractions.putWithUniqueKey(key, content, attributes);
    }
}

File: src/main/java/dev/kensa/state/CapturedInteractionBuilder.kt
package dev.kensa.state

import dev.kensa.render.diagram.directive.ArrowStyle
import dev.kensa.util.Attributes
import dev.kensa.util.Attributes.Companion.of
import dev.kensa.util.Attributes.Key.Arrow
import dev.kensa.util.Attributes.Key.Group
import java.time.Instant

class CapturedInteractionBuilder private constructor(private val fromParty: Party) {
    private var attributes = of()
    private var toParty: Party? = null
    private var group: String? = null
    private var content: Any? = null
    private var contentDescriptor: String? = null
    private var arrowStyle : ArrowStyle = ArrowStyle.UmlSynchronous
    private var timestamp : Long? = null

    fun arrowStyle(arrowStyle: ArrowStyle) = apply { this.arrowStyle = arrowStyle }

    fun to(toParty: Party): CapturedInteractionBuilder = apply { this.toParty = toParty }

    fun group(group: String): CapturedInteractionBuilder = apply { this.group = group }

    fun underTest(isUnderTest: Boolean): CapturedInteractionBuilder = apply { if (group == null && isUnderTest) group = "Test" }

    fun with(timestamp: Instant) = with(timestamp.toEpochMilli())

    fun with(timestamp: Long) = apply { this.timestamp = timestamp }

    fun with(content: Any, contentDescriptor: String): CapturedInteractionBuilder = apply {
        this.content = content
        this.contentDescriptor = contentDescriptor
    }

    fun with(attributes: Attributes): CapturedInteractionBuilder = apply { this.attributes = attributes }

    fun applyTo(interactions: CapturedInteractions) {
        attributes = group?.let { attributes.merge(of(Group, it, Arrow, arrowStyle)) }
            ?: attributes.merge(of(Arrow, arrowStyle))

        val key = "$contentDescriptor __idx __from ${fromParty.asString()} to ${toParty!!.asString()}"

        interactions.putWithUniqueKey(key, content, timestamp ?: System.currentTimeMillis(), attributes)
    }

    companion object {
        @JvmStatic
        fun from(party: Party): CapturedInteractionBuilder = CapturedInteractionBuilder(party)
    }
}